File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {
  private int capacity;
  private Map<Integer, Integer> cache;

  public LRUCache(int capacity) {
    this.cache =
        new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true) {
          @Override
          protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
            return size() > capacity;
          }
        };
  }

  public int get(int key) {
    return cache.getOrDefault(key, -1);
  }

  public void put(int key, int value) {
    cache.put(key, value);
  }
}

/**
 * Your LRUCache object will be instantiated and called as such: LRUCache obj = new
 * LRUCache(capacity); int param_1 = obj.get(key); obj.put(key,value);
 */

File: leetcode/Medium/0146-lru-cache/0146-lru-cache.py
class LRUCache:
    def __init__(self, capacity: int):
        self.cache = dict()
        self.capacity = capacity

    def get(self, key: int) -> int:
        value = self.cache.get(key, None)
        if value is not None:
            del self.cache[key]
            self.cache[key] = value
            return value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            del self.cache[key]
        self.cache[key] = value
        if len(self.cache) > self.capacity:
            first_item_key = next(iter(self.cache))
            del self.cache[first_item_key]


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
